digi/xbee/xsocket.py

```diff
@@ -0,0 +1,814 @@
+            ValueError: if ``ip_protocol`` is ``None``.
+            XBeeException: if the connection with the XBee device is not open.
+        """
+        if xbee_device is None:
+            raise ValueError("XBee device cannot be None")
+        if not isinstance(xbee_device, CellularDevice):
+            raise ValueError("XBee device must be a Cellular device")
+        if ip_protocol is None:
+            raise ValueError("IP protocol cannot be None")
+        if not xbee_device.is_open():
+            raise XBeeException("XBee device must be open")
+
+        # Initialize internal vars.
+        self.__xbee_device = xbee_device
+        self.__ip_protocol = ip_protocol
+        self.__socket_id = None
+        self.__connected = False
+        self.__source_port = None
+        self.__is_listening = False
+        self.__backlog = None
+        self.__timeout = self.__DEFAULT_TIMEOUT
+        self.__data_received = bytearray()
+        self.__data_received_lock = threading.Lock()
+        self.__data_received_from_dict = OrderedDict()
+        self.__data_received_from_dict_lock = threading.Lock()
+        # Initialize socket callbacks.
+        self.__socket_state_callback = None
+        self.__data_received_callback = None
+        self.__data_received_from_callback = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def connect(self, address):
+        """
+        Connects to a remote socket at the given address.
+
+        Args:
+            address (Tuple): A pair ``(host, port)`` where ``host`` is the domain name or string representation of an
+                IPv4 and ``port`` is the numeric port value.
+
+        Raises:
+            TimeoutException: if the connect response is not received in the configured timeout.
+            ValueError: if ``address`` is ``None`` or not a pair ``(host, port)``.
+            ValueError: if ``port`` is less than 1 or greater than 65535.
+            XBeeException: if the connection with the XBee device is not open.
+            XBeeSocketException: if the connect status is not ``SUCCESS``.
+        """
+        # Check address and its contents.
+        if address is None or len(address) != 2:
+            raise ValueError("Invalid address, it must be a pair (host, port).")
+
+        host = address[0]
+        port = address[1]
+        if isinstance(host, IPv4Address):
+            host = str(host)
+        if port < 1 or port > 65535:
+            raise ValueError("Port number must be between 1 and 65535.")
+
+        # If the socket is not created, create it first.
+        if self.__socket_id is None:
+            self.__create_socket()
+
+        lock = threading.Condition()
+        received_state = list()
+
+        # Define the socket state received callback.
+        def socket_state_received_callback(socket_id, state):
```

Comment:
Question, a method defined inside another method is only accessible from the parent method, right? This callback is not actually public.
Also, I am assuming there is a reason for defining it here, is it a best practices thing that Python developers recommend?